pkg/task/storage.go

```diff
@@ -0,0 +1,262 @@
+// 3. the worker pops a task of the heap
+// 4. testground is restarted or crashes in the middle of the work.
+// 5. testground (restarted) will add this task to the heap again regardless of the max length constraint
+//
+// In order to keep the storage requirements relatively small, older keys are evicted.
+type Queue struct {
+	sync.Mutex
+	tq *taskQueue  // priority task queue
+	db *leveldb.DB // on-disk key-value databse
+	eo []*evict    // eviction order when there are too many keys.
+
+	max     int              // the maximum number of tasks to keep in the databse
+	onEvict EvictionFunction // Additional cleanup function when eviction occurs.
+}
+
+// Add an item to the priority queue
+// 1. Check if there are more than the maximum allowed keys in the database
+//    a. if there are, evict old keys
+//    b. call eviction function
+// 2. Persist the new task to the database
+// 3. Push the new task onto the queue
+func (q *Queue) Push(tsk *Task) error {
+	q.Lock()
+	defer q.Unlock()
+
+	// special case: there are too many items enqueued already. can't push; try again later.
+	if q.tq.Len() >= q.max {
+		return ErrQueueFull
+	}
+	// evict keys from the database until we have less than the max.
+	// we have one evict per key in q.db, so the len(q.eo) is the number of keys in leveldb.
+	for keys := len(q.eo); keys >= q.max; keys-- {
+		key := q.eo[0].Key
+		err := q.db.Delete([]byte(key), &opt.WriteOptions{
+			Sync: true,
+		})
+		if err != nil {
+			return err
+		}
+		// Cleanup
+		q.onEvict(key)
+		q.eo = q.eo[1:]
+	}
+
+	// Persist this task to the database
+	err := q.put(tsk)
+	if err != nil {
+		return err
+	}
+	// Add this task to the eviction order
+	q.eo = append(q.eo, &evict{tsk.ID, tsk.Created})
+	// Push this task to the queue
+	heap.Push(q.tq, tsk)
+	return nil
+}
+
+// Get an Task from the K-V store
+// 1. Lookup the key in the database.
+// 2. Unmarshal the task into the provided task pointer.
+func (q *Queue) Get(id string, tsk *Task) error {
+	key := []byte(id)
+	val, err := q.db.Get(key, nil)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(val, tsk)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// unexported; put value into the K-V store.
+func (q *Queue) put(tsk *Task) error {
+	key := []byte(tsk.ID)
+	val, err := json.Marshal(tsk)
+	if err != nil {
+		return err
+	}
+	return q.db.Put(key, val, &opt.WriteOptions{
+		Sync: true,
+	})
+}
+
+// get the next item from the priority queue
+// Pop the task off of the queue
+// The task remains in the database, but is no longer in the heap.
+// As the state of the task changes (i.e. to mark the task completed, use SetTaskState)
+func (q *Queue) Pop() (*Task, error) {
+	q.Lock()
+	defer q.Unlock()
+	if q.tq.Len() == 0 {
+		return nil, ErrQueueEmpty
+	}
+	tsk := heap.Pop(q.tq).(*Task)
+	return tsk, nil
+}
+
+// Change the state of a task in the K-V store
+// This method
+func (q *Queue) AppendTaskState(id string, state TaskState) error {
```

Comment:
@nonsense @raulk I am about *this close* to changing this method to being a method of *Task instead. The only reason I haven't done it is it would add a pointer to leveldb into the task struct which would sometimes be nil.

Currently, "Push" and "Pop" operate on *Task. Obviously, a Task being pushed onto the queue does not have a reference to leveldb, and I want to avoid having any half-functional Tasks anywhere.

Right now, `Task` is just a data model without methods, and I think this is a feature. The `Task` is the same object on client and server, in the database and on the wire. We could change this model out for apache thrift, a protobuf, avro or any other serialization technique without having to change anything else.

Should I try to make Queue stick to being a typical work Queue and move this method somewhere else or maybe change Queue to be called `TaskMaster` and embrase the part-queue-part-key-value nature of this thing.